proposals/extensions.md

```diff
@@ -1219,34 +1219,68 @@ implicit extension E for int
 
 # Implementation details
 
-TODO3(static) revise this to use a regular struct
+Extensions are emitted as structs with an extension marker method and an instance field for non-static extensions.  
+The type is marked with Obsolete and CompilerFeatureRequired attributes.  TODO we'll need to relax that
 
-struct Extension
-{
-  UnderlyingType underlyingValue;
-}
-// Avoid copying via Unsafe.As 
+## Marker method
 
-Extensions are implemented as ref structs with an extension marker method.  
-The type is marked with Obsolete and CompilerFeatureRequired attributes.  
 The extension marker method encodes the underlying type and base extensions as parameters in that order.  
-The marker method is called `<ImplicitExtension>$` for implicit extensions and 
+It is private and static, and is called `<ImplicitExtension>$` for implicit extensions and 
 `<ExplicitExtension>$` for explicit extensions.  
+
 For example: `implicit extension R for UnderlyingType : BaseExtension1, BaseExtension2` yields
 `private static void <ImplicitExtension>$(UnderlyingType, BaseExtension1, BaseExtension2)`.  
 
-If the extension has any instance member, then we emit a ref field (of underlying type)
-into the ref struct and a constructor.  
-TODO2(instance) The wrapping can be done with a static unspeakable factory method  
+## Instance field
 
-Values of extension types are left as values of the underlying value type, until an extension
-member is accessed. When an extension member is accessed, an extension instance is created
-with a reference to the underlying value and the member is accessed on that instance.
+If the extension not static, then we emit a private instance field of the underlying type
+into the struct.  
+We use the default layout for structs (sequential layout, with pack and size 0)
+ensuring that the single instance field is placed at offset zero.
 
+Note: Although we could not find an explicit statement to that effect,
+this behavior falls out from ECMA 335 (II.10.1.2), which states for sequential layout:
```

Comment:
The sequential/explicit/auto type layout attributes describe unmanaged layout of the type (ie what you get when you marshal that type using built-in runtime marshaller. If the type is blittable, the managed layout of the type is going to match that unmanaged. If the type is not blittable, the runtime is free to use any layout it wants for the type.

For example, consider the following struct. It is non-blittable structs. The comments have offsets assigned to the fields by the current CoreCLR type loader. You can see that it is anything but sequential:
```
struct MyStruct
{
    public int a; // offset 16
    public object b; // offset 0
    public int c; // offset 20
    public object d; // offset 8
}
```

Reply:
(If you have a non-blittable sequential struct with a single field, I believe that all current runtimes are going to place that field at offset 0 in the managed view. It do not think it is written down anywhere. )

Reply:
It might be good to just guarantee this specific case (1 field in a sequential struct) in the augments.